### PR TITLE
Fabric memory stats

### PR DESCRIPTION
**Pull request by sbekele81**

### CODE REVIEW

File: xprof/xprof.rb.in [hunk 1]
```diff
@@ -385,16 +385,11 @@ def env_tracers
...
   # Sample
-  #   Currently the same `so` does the tracing, and the sampling
-  #   This mean that is the local rank is not part of the `traced-ranks`
-  #   No sampling will be performed
-  if sampling?
+  if sampling? 
     LOGGER.debug('Sampling Enabled')
     h['LTTNG_UST_SAMPLING'] = 1
     h['LTTNG_UST_SAMPLING_ENERGY'] = 1
-    # The current only reliable way to use zes api
```
Comment (TApplencourt) [inline xprof/xprof.rb.in:395]: Why did you remove my comment? :'0 :')

File: xprof/btx_timeline.cpp [hunk 2]
```diff
@@ -72,78 +78,122 @@ static perfetto_uuid_t get_parent_counter_track_uuid(timeline_dispatch_t *dispat
   auto *process = track_descriptor->mutable_process();
   process->set_pid(hp_uuid);
   std::ostringstream oss;
-  oss << "Hostname " << hostname << " | Process " << process_id << " | Device " << did;
+  oss << "Hostname " << hostname << " | Device " << deviceIdx;
```
Comment (TApplencourt) [inline xprof/btx_timeline.cpp:81]: Do we need that?
I would like this PR to not change the default non sampling timeline.

File: xprof/btx_timeline.cpp [hunk 3]
```diff
@@ -72,78 +78,122 @@ static perfetto_uuid_t get_parent_counter_track_uuid(timeline_dispatch_t *dispat
...
   process->set_process_name(oss.str());
   return hp_uuid;
 }
 
-static perfetto_uuid_t
-get_counter_track_uuuid(timeline_dispatch_t *dispatch,
-                        std::unordered_map<hp_ddomain_t, perfetto_uuid_t> &counter_tracks,
-                        const std::string track_name, std::string hostname, uint64_t process_id,
-                        thapi_device_id did, thapi_domain_idx domain, int64_t unit_multiplier = 1) {
+static perfetto_uuid_t get_counter_track_uuuid(timeline_dispatch_t *dispatch,
```
Comment (TApplencourt) [inline xprof/btx_timeline.cpp:87]: why this code changed? What is details?

File: ze/tracer_ze_helpers.include.c [hunk 4]
```diff
@@ -838,6 +834,17 @@ static void intializeFrequency() {
         _sampling_freqDomainCounts[driverIdx][deviceIdx] = 0;
         free(_sampling_hFrequencies[driverIdx][deviceIdx]);
       }
+      for (uint32_t domainIdx = 0; domainIdx <  _sampling_freqDomainCounts[driverIdx][deviceIdx]; domainIdx++) {
+        zes_freq_properties_t freqProps ={0};
```
Comment (TApplencourt) [inline ze/tracer_ze_helpers.include.c:838]: Please run clang-format or fix the intendation

File: xprof/btx_timeline.cpp (hunk 2 above)
Comment (sbekele81) [reply xprof/btx_timeline.cpp:81]: This is not non sampling timeline. We were printing process_id but it is always 0 for sampling. The non sampling function still uses process_id and safe :) .

Comment (TApplencourt) [reply xprof/btx_timeline.cpp:81]: What do you mean `get_parent_counter_track_uuid` is not used by the regular timeline? Only by the sampling?

File: xprof/btx_timeline.cpp [hunk 5]
```diff
@@ -5,7 +5,7 @@
 #include <iomanip>  // set precision
 #include <iostream> // stdcout
 #include <map>
-#include <optional>
+#include <string>
```
Comment (TApplencourt) [inline xprof/btx_timeline.cpp:8]: but you are still using optonal

File: xprof/btx_timeline.cpp [hunk 6]
```diff
@@ -30,17 +30,23 @@ struct timeline_dispatch_s {
...
-  std::unordered_map<hp_dsdev_t, perfetto_uuid_t> hp_dsdev2cpytracks;
-
+  std::unordered_map<h_device_t, perfetto_uuid_t> hp_device2countertracks;
+  std::unordered_map<hp_ddomain_t, perfetto_uuid_t> hp_ddomain2telmtracks;
+  std::unordered_map<hp_ddomain_t, perfetto_uuid_t> hp_ddomain2cpytracks;
+  std::unordered_map<hp_dfsdev_t, perfetto_uuid_t> hp_dfsdev2fptracks;
   perfetto_pruned::Trace trace;
 };
+
+struct Details {
```
Comment (TApplencourt) [inline xprof/btx_timeline.cpp:40]: Please find a better name.
Always can it just be a `tuple`? We use tuples everywhere.

File: xprof/btx_timeline.cpp (hunk 2 above)
Comment (TApplencourt) [reply xprof/btx_timeline.cpp:81]: Oh good call. Missed the `counter` in the name! Thanks

File: xprof/btx_timeline.cpp [hunk 7]
```diff
@@ -72,78 +78,122 @@ static perfetto_uuid_t get_parent_counter_track_uuid(timeline_dispatch_t *dispat
...
+    oss << track_name << " Module " << domain;
+    oss <<  " | " <<(details->RxTx ? "WR BW" : "RD BW");
+  }
+  else if (track_name=="Allocated Memory (%)") {
+    oss << track_name << " Module " << domain;
+  }
+  else if (track_name==" CopyEngine (%)" || track_name==" ComputeEngine (%)") {
+    oss << track_name << " | SubDevice " << domain;
+  }
+  else if (track_name=="   Power") {
```
Comment (TApplencourt) [inline xprof/btx_timeline.cpp:138]: Why not only one space?

File: xprof/btx_timeline.cpp [hunk 8]
```diff
@@ -72,78 +78,122 @@ static perfetto_uuid_t get_parent_counter_track_uuid(timeline_dispatch_t *dispat
...
-  // Create new track
   auto *packet = dispatch->trace.add_packet();
   packet->set_timestamp(0);
   packet->set_trusted_packet_sequence_id(TRUSTED_PACKED_SEQUENCE_ID);
   auto *track_descriptor = packet->mutable_track_descriptor();
   track_descriptor->set_uuid(hp_dev_uuid);
   track_descriptor->set_parent_uuid(hp_uuid);
   std::ostringstream oss;
-  oss << track_name << " | Domain " << domain;
+  if (track_name=="FabricT" && details) {
```
Comment (TApplencourt) [inline xprof/btx_timeline.cpp:124]: So we cannot ever use `FabricT` without `details`?

File: xprof/btx_timeline.cpp [hunk 9]
```diff
@@ -72,78 +78,122 @@ static perfetto_uuid_t get_parent_counter_track_uuid(timeline_dispatch_t *dispat
...
-                                             uint64_t process_id, thapi_device_id did,
-                                             thapi_domain_idx domain) {
-  // Extra leading space in the name field to make GPU Power the first track
-  return get_counter_track_uuuid(dispatch, dispatch->hp_ddomain2pwrtracks, "  GPU Power", hostname,
-                                 process_id, did, domain);
+
+
+static perfetto_uuid_t get_frequency_track_uuuid(timeline_dispatch_t *dispatch, const std::string &hostname, uint64_t process_id,
+                                                 uint64_t did, uint32_t deviceIdx, uint64_t hFrequency, uint32_t subDevice,  std::optional<Details> options) {
+     return get_counter_track_uuuid(dispatch, dispatch->hp_ddomain2telmtracks, "  Ferquency", hostname, process_id, did, deviceIdx, hFrequency, subDevice);
```
Comment (TApplencourt) [inline xprof/btx_timeline.cpp:176]: Frequency

File: xprof/btx_timeline.cpp [hunk 10]
```diff
@@ -72,78 +78,122 @@ static perfetto_uuid_t get_parent_counter_track_uuid(timeline_dispatch_t *dispat
...
 }
 
-static perfetto_uuid_t get_computeEU_track_uuuid(timeline_dispatch_t *dispatch,
-                                                 std::string hostname, uint64_t process_id,
-                                                 thapi_device_id did, thapi_sdevice_idx subDevice) {
-  return get_counter_track_uuuid(dispatch, dispatch->hp_dsdev2cpetracks, "ComputeEngine (%)",
-                                 hostname, process_id, did, subDevice, 100);
+static perfetto_uuid_t get_Bandwidth_track_uuuid(timeline_dispatch_t *dispatch, const std::string &hostname, uint64_t process_id,
```
Comment (TApplencourt) [inline xprof/btx_timeline.cpp:179]: no capital `B` on `Bandwidt` please

File: xprof/btx_timeline.cpp [hunk 11]
```diff
@@ -153,31 +203,59 @@ static void add_event_DTelemetry(timeline_dispatch_t *dispatch, std::string host
...
+                                 uint64_t process_id, uint64_t thread_id, uint64_t did, uint32_t deviceIdx, uintptr_t hMemModule, uint32_t subDevice, uint64_t timestamp, float pBandwidth, float rdBandwidth, float wtBandwidth, float occupancy) {
+  // Define details for RX throughput.
+  Details details = {false, 0, 0};
+  add_event_DTelemetry(dispatch, hostname, process_id, thread_id, did, deviceIdx, hMemModule, subDevice, timestamp,
+                       rdBandwidth, get_Bandwidth_track_uuuid, "Memory Read BW", details);
+  details.RxTx = true;
+  add_event_DTelemetry(dispatch, hostname, process_id, thread_id, did, deviceIdx, hMemModule, subDevice, timestamp,
+                       wtBandwidth, get_Bandwidth_track_uuuid, "Memory Write BW", details);
+  add_event_DTelemetry(dispatch, hostname, process_id, thread_id, did, deviceIdx, hMemModule, subDevice, timestamp,
+                       occupancy, get_Occupancy_track_uuuid, "Memory Occupancy", details);
```
Comment (TApplencourt) [inline xprof/btx_timeline.cpp:216]: Please avoid mutation.
Use scope and re-instanciate a `Details details`.
Also `{true, 0, 0}` sound like a code smell to me.  DoesmemModule relly have fabric and remorp port id?

File: xprof/btx_timeline.cpp [hunk 12]
```diff
@@ -153,31 +203,59 @@ static void add_event_DTelemetry(timeline_dispatch_t *dispatch, std::string host
...
+                                 uint64_t process_id, uint64_t thread_id, uint64_t did, uint32_t deviceIdx, uintptr_t hFabricPort,
+                                 uint32_t subDevice, uint64_t timestamp, uint32_t fabricId, uint32_t remotePortId, 
+                                 float rxThroughput, float txThroughput, float rxSpeed, float txSpeed) {
+  // Define details for RX throughput.
+  Details details = {false, fabricId, remotePortId};
+  add_event_DTelemetry(dispatch, hostname, process_id, thread_id, did, deviceIdx, hFabricPort, subDevice, timestamp,
+                       rxThroughput, get_fpThroughput_track_uuuid, "Fabric ThroughputRX", details);
+
+  details.RxTx = true;
+  add_event_DTelemetry(dispatch, hostname, process_id, thread_id, did, deviceIdx, hFabricPort, subDevice, timestamp,
```
Comment (TApplencourt) [inline xprof/btx_timeline.cpp:230]: Please add some documentation somewhere to explain `details`. What information he has and how it's useful it it's.

File: xprof/xprof.rb.in (hunk 1 above)
Comment (sbekele81) [reply xprof/xprof.rb.in:395]: Comment included :)

File: xprof/btx_timeline.cpp (hunk 5 above)
Comment (sbekele81) [reply xprof/btx_timeline.cpp]: It is already included in xprof_utils.hpp.

File: xprof/btx_timeline.cpp (hunk 7 above)
Comment (sbekele81) [reply xprof/btx_timeline.cpp]: If you have more leading space you would be the first. If power takes one space, the others' order will alphabetical and would lose the order. I agree I need more robust way to do it.

File: xprof/btx_timeline.cpp (hunk 8 above)
Comment (sbekele81) [reply xprof/btx_timeline.cpp]: Yes, in our context we are using details for Fabric and Memory stats.  We wanted not to lose the remote and the fabric Ids on the timeline. In addition, the direction of the data movement. In the case of memory stats we need the direction of the movement. These are not available on the other metrics so, to avoid  redundancy, I used details as optional for the two.

File: xprof/xprof.rb.in (hunk 1 above)
Comment (TApplencourt) [reply xprof/xprof.rb.in:395]: Thanks!

File: xprof/xprof.rb.in [hunk 13]
```diff
@@ -388,13 +388,13 @@ def env_tracers
   #   Currently the same `so` does the tracing, and the sampling
   #   This mean that is the local rank is not part of the `traced-ranks`
   #   No sampling will be performed
-  if sampling?
+  if sampling?
```
Comment (TApplencourt) [inline xprof/xprof.rb.in:391]:
```suggestion
  if sampling?
```

File: xprof/xprof.rb.in [hunk 14]
```diff
@@ -388,13 +388,13 @@ def env_tracers
...
-  if sampling?
+  if sampling? 
     LOGGER.debug('Sampling Enabled')
     h['LTTNG_UST_SAMPLING'] = 1
     h['LTTNG_UST_SAMPLING_ENERGY'] = 1
     # The current only reliable way to use zes api
-    #   is to call zesInit and set ZES_ENABLE_SYSMAN to 0
-    h['ZES_ENABLE_SYSMAN'] = 0 if OPTIONS[:'backend-names'].include?('ze')
+    #  is to call zesInit and set ZES_ENABLE_SYSMAN to 0
+    h['ZES_ENABLE_SYSMAN'] = 0  if OPTIONS[:'backend-names'].include?('ze')
```
Comment (TApplencourt) [inline xprof/xprof.rb.in:397]:
```suggestion
    h['ZES_ENABLE_SYSMAN'] = 0 if OPTIONS[:'backend-names'].include?('ze')
```

File: ze/btx_zeinterval_callbacks.cpp [hunk 15]
```diff
@@ -765,87 +765,281 @@ static void zeEventDestroy_exit_callback(void *btx_handle, void *usr_data, int64
...
+      hash *= 0x100000001b3; // FNV prime
+    }
 
-  // Watt conversion
-  btx_push_message_lttng_power(btx_handle, hostname, 0, 0, prev_ts, BACKEND_ZE, (uint64_t)hDevice,
-                               (thapi_domain_idx)domain,
-                               (energy - prev_energy) / (double)(sampling_ts - prev_sampling_ts));
+    return {hash, deviceIdx};
+  }
+  return {0, 0}; // Return 0 values if not found
```
Comment (TApplencourt) [inline ze/btx_zeinterval_callbacks.cpp:783]: I guess we should use optional to be cleaner.

File: ze/btx_zeinterval_callbacks.cpp [hunk 16]
```diff
@@ -765,87 +765,281 @@ static void zeEventDestroy_exit_callback(void *btx_handle, void *usr_data, int64
...
+    size_t _pFabricPortThroughput_val_length,
+    zes_fabric_port_throughput_t *pFabricPortThroughput_val) {
+  auto *data = static_cast<data_t *>(usr_data);
+  const auto it0 = data->fabricPort_property.find({hostname, vpid, hDevice, hFabricPort});
+  if (it0 != data->fabricPort_property.cend()) {
+    // Get fabricPort properties: subdevice ID, fabeicId ...
+    auto subDevice = it0->second.subdeviceId;
+    auto fabricId = it0->second.portId.fabricId;
+    auto remotePortId = pFabricPortState_val->remotePortId.fabricId;
+    // Current Speed (not used currently in the timeline)
```
Comment (TApplencourt) [inline ze/btx_zeinterval_callbacks.cpp:799]: Please explain why divided by `8` I guess bits -> Bytes?

File: xprof/btx_timeline.cpp [hunk 17]
```diff
@@ -72,78 +79,146 @@ static perfetto_uuid_t get_parent_counter_track_uuid(timeline_dispatch_t *dispat
...
-                        std::unordered_map<hp_ddomain_t, perfetto_uuid_t> &counter_tracks,
-                        const std::string track_name, std::string hostname, uint64_t process_id,
-                        thapi_device_id did, thapi_domain_idx domain, int64_t unit_multiplier = 1) {
+static perfetto_uuid_t get_counter_track_uuuid(
+    timeline_dispatch_t *dispatch,
+    std::unordered_map<hp_ddomain_t, perfetto_uuid_t> &counter_tracks,
+    const std::string &track_name, const std::string &hostname, uint64_t process_id,
+    thapi_device_id did, uint32_t deviceIdx, uint64_t tHandle, thapi_domain_idx domain,
+    std::optional<Details> details = std::nullopt,
+    std::unordered_map<hp_dfsdev_t, perfetto_uuid_t> *counter_tracks_fp = nullptr) {
```
Comment (TApplencourt) [inline xprof/btx_timeline.cpp:94]: Does a `std::optional` of a reference would make more sense?

File: xprof/btx_timeline.cpp [hunk 18]
```diff
@@ -72,78 +79,146 @@ static perfetto_uuid_t get_parent_counter_track_uuid(timeline_dispatch_t *dispat
...
+  } else {
+    auto [it, inserted] =
+        counter_tracks.insert({{hostname, process_id, did, domain, tHandle}, hp_dev_uuid});
+    auto &potential_uuid = it->second;
+    if (!inserted)
+      return potential_uuid;
+
+    hp_uuid = get_parent_counter_track_uuid(dispatch, hostname, process_id, did, deviceIdx);
+    hp_dev_uuid = gen_perfetto_uuid();
+    potential_uuid = hp_dev_uuid;
```
Comment (TApplencourt) [inline xprof/btx_timeline.cpp:117]: Look like those instruction are the same in both branch.  I think it make sense to remove it outside of the if-else

File: xprof/btx_timeline.cpp [hunk 19]
```diff
@@ -72,78 +79,146 @@ static perfetto_uuid_t get_parent_counter_track_uuid(timeline_dispatch_t *dispat
...
+  } else {
+    auto [it, inserted] =
+        counter_tracks.insert({{hostname, process_id, did, domain, tHandle}, hp_dev_uuid});
```
Comment (TApplencourt) [inline xprof/btx_timeline.cpp:110]: Look like the only difference is that one use `counter_tracks` or `counter_tracks_fp` and then insert either `details->RxTx` or `tHandle`, can we just template `get_counter_track_uuuid` then?

File: ze/btx_zeinterval_callbacks.cpp [hunk 20]
```diff
@@ -765,87 +765,281 @@ static void zeEventDestroy_exit_callback(void *btx_handle, void *usr_data, int64
...
+        activeTime = static_cast<double>((UINT64_MAX - prev_engineStats.activeTime) +
+                                         pEngineStats_val->activeTime + 1) *
+                     100 / time_diff;
+      DeviceHash uuid_idx = get_device_hash(usr_data, hostname, vpid, hDevice);
+      if (engineProps.type == ZES_ENGINE_GROUP_COMPUTE_ALL) {
+        btx_push_message_lttng_computeEU(btx_handle, hostname, 0, 0, prev_ts, BACKEND_ZE,
+                                         uuid_idx.hash, uuid_idx.deviceIdx, (uint64_t)hEngine,
+                                         subDevice, int(activeTime));
+      }
+      if (engineProps.type == ZES_ENGINE_GROUP_COPY_ALL) {
```
Comment (TApplencourt) [inline ze/btx_zeinterval_callbacks.cpp:934]: Should be a `il else` to showcase that both branch cannot be taken

File: ze/btx_zeinterval_callbacks.cpp [hunk 21]
```diff
@@ -765,87 +765,281 @@ static void zeEventDestroy_exit_callback(void *btx_handle, void *usr_data, int64
...
+        activeTime =
+            static_cast<double>(pEngineStats_val->activeTime - prev_engineStats.activeTime) * 100 /
+            time_diff;
+      else
+        activeTime = static_cast<double>((UINT64_MAX - prev_engineStats.activeTime) +
+                                         pEngineStats_val->activeTime + 1) *
+                     100 / time_diff;
+      DeviceHash uuid_idx = get_device_hash(usr_data, hostname, vpid, hDevice);
+      if (engineProps.type == ZES_ENGINE_GROUP_COMPUTE_ALL) {
+        btx_push_message_lttng_computeEU(btx_handle, hostname, 0, 0, prev_ts, BACKEND_ZE,
```
Comment (TApplencourt) [inline ze/btx_zeinterval_callbacks.cpp:930]: Why do we need the `0,0`. If it's always `0,0` we should remove it from the schema

File: ze/btx_zeinterval_callbacks.cpp (hunk 16 above)
Comment (sbekele81) [reply ze/btx_zeinterval_callbacks.cpp:799]: yes i think I have to comment it.

File: xprof/btx_timeline.cpp (hunk 9 above)
Comment (sbekele81) [reply xprof/btx_timeline.cpp]: Thanks, corrected.

File: ze/btx_zeinterval_callbacks.cpp (hunk 21 above)
Comment (sbekele81) [reply ze/btx_zeinterval_callbacks.cpp]: Now, we removed that and use separate stream for sampling

File: xprof/btx_timeline.cpp (hunk 19 above)
Comment (sbekele81) [reply xprof/btx_timeline.cpp]: Used template.

File: xprof/btx_timeline.cpp (hunk 6 above)
Comment (sbekele81) [reply xprof/btx_timeline.cpp]: Changed to tuple.

File: ze/btx_zeinterval_callbacks.cpp (hunk 15 above)
Comment (TApplencourt) [reply ze/btx_zeinterval_callbacks.cpp:783]: (I don't understand why it was  put as resolved)

File: ze/btx_zeinterval_callbacks.cpp [hunk 22]
```diff
@@ -765,87 +765,281 @@ static void zeEventDestroy_exit_callback(void *btx_handle, void *usr_data, int64
...
+}
+
+static void lttng_ust_ze_sampling_fabricPortProperties_callback(
+    void *btx_handle, void *usr_data, int64_t ts, const char *hostname, int64_t vpid, uint64_t vtid,
+    ze_device_handle_t hDevice, zes_fabric_port_handle_t hFabricPort,
+    size_t _pFabricPortProperties_val_length,
+    zes_fabric_port_properties_t *pFabricPortProperties_val) {
+  auto *data = static_cast<data_t *>(usr_data);
+  data->fabricPort_property[{hostname, vpid, (ze_device_handle_t)hDevice,
+                             (zes_fabric_port_handle_t)hFabricPort}] = *pFabricPortProperties_val;
```
Comment (TApplencourt) [inline ze/btx_zeinterval_callbacks.cpp:1006]: Why the cast? `hFabricPort` is already a `zes_fabric_port_handle_t`

File: ze/btx_zeinterval_callbacks.hpp [hunk 23]
```diff
@@ -51,6 +69,12 @@ using btx_event_desct_t =
 using btx_command_list_desc_t =
     std::tuple<ze_command_queue_desc_t, ze_device_handle_t, bool /*hCommandListIsImmediate*/>;
 
+struct DeviceHash {
+    uint64_t hash;
+    uint32_t deviceIdx;
+};
```
Comment (TApplencourt) [inline ze/btx_zeinterval_callbacks.hpp:75]: If you never mutate member of `DeviceHash`I would prefer to use tuple just because we use them everywhere else

File: ze/btx_zeinterval_callbacks.cpp [hunk 24]
```diff
@@ -765,87 +765,281 @@ static void zeEventDestroy_exit_callback(void *btx_handle, void *usr_data, int64
...
+      double activeTime = 0;
+      if (pEngineStats_val->activeTime > prev_engineStats.activeTime)
+        activeTime =
+            static_cast<double>(pEngineStats_val->activeTime - prev_engineStats.activeTime) * 100 /
+            time_diff;
+      else
+        activeTime = static_cast<double>((UINT64_MAX - prev_engineStats.activeTime) +
+                                         pEngineStats_val->activeTime + 1) *
+                     100 / time_diff;
+      DeviceHash uuid_idx = get_device_hash(usr_data, hostname, vpid, hDevice);
```
Comment (TApplencourt) [inline ze/btx_zeinterval_callbacks.cpp:931]: You don't test tht this guy doesn't return `{0,0}` (or optional not set)
Do you want to crash or to not send an message in this case?

File: ze/btx_zeinterval_callbacks.cpp (hunk 15 above)
Comment (sbekele81) [reply ze/btx_zeinterval_callbacks.cpp]: Sorry, it was mixed up. I used optional with a message to handle not found.

File: ze/btx_zeinterval_callbacks.cpp (hunk 24 above)
Comment (sbekele81) [reply ze/btx_zeinterval_callbacks.cpp]: I used optional and a message if null value returned.

File: ze/btx_zeinterval_callbacks.hpp (hunk 23 above)
Comment (sbekele81) [reply ze/btx_zeinterval_callbacks.hpp]: Used tuple instead.

File: ze/btx_zeinterval_callbacks.cpp (hunk 22 above)
Comment (sbekele81) [reply ze/btx_zeinterval_callbacks.cpp]: Removed, thanks.

File: ze/btx_zeinterval_callbacks.cpp [hunk 25]
```diff
@@ -765,87 +765,299 @@ static void zeEventDestroy_exit_callback(void *btx_handle, void *usr_data, int64
...
+  }
+  return std::nullopt; // Return empty optional if not found
 }
 
-static void lttng_ust_ze_sampling_gpu_frequency_callback(void *btx_handle, void *usr_data,
-                                                         int64_t ts, const char *hostname,
-                                                         int64_t vpid, uint64_t vtid,
-                                                         ze_device_handle_t hDevice,
-                                                         uint32_t domain, uint64_t frequency) {
+uint64_t calculate_delta(uint64_t current_val, uint64_t prev_val) {
```
Comment (TApplencourt) [inline ze/btx_zeinterval_callbacks.cpp:786]: Do you need to return `prev_val-current_val` if not equal?  If not then this function is maybe better called `check_delta` or something like that

File: ze/btx_zeinterval_callbacks.cpp [hunk 26]
```diff
@@ -765,87 +765,299 @@ static void zeEventDestroy_exit_callback(void *btx_handle, void *usr_data, int64
...
+  // return zero if equal
+  if (current_val == prev_val) {
+    return 0;
+  }
```
Comment (TApplencourt) [inline ze/btx_zeinterval_callbacks.cpp:790]: You don't need it I think I `current_val == prev_val` then `current_val >= prev_val` will be true and ` return current_val - prev_val;` will be 0

Comment (sbekele81) [reply ze/btx_zeinterval_callbacks.cpp:790]: Yes, I can handle the divide by zero in the calling function (for the timestamp).

**Review by TApplencourt — APPROVED** [review]
